Migrate Administration Page to Symfony

mickaelandrieu opened the pull request:
<!-- Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "AnsAwers" rows with the necessary information: -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Self described
| Type?         | improvement
| Category?     | BO
| BC breaks?    | yes (controller override)
| Deprecations? | probably
| Fixed ticket? | [BOGOSS-49](http://forge.prestashop.com/browse/BOGOSS-49)
| How to test?  | Access `Configure > Advanced Parameters > Administration`

Todo:
- [x] Deletions
  - [x] Remove `AdminPreferencesController`
  - [x] Remove related old templates
- [x] Creations
  - [x] `PrestaShopBundle/Controller/AdvancedParameters/AdministrationController`
  - [x] Add routing
  - [x] Complete `Link` class to map PrestaShop menu to the new page
  - [x] Create `General`, `UploadQuota` and `Notifications` Symfony forms
  - [x] Added help/info tips labels and messages
  - [x] ~~Create the Twig template using PrestaShop UI Kit~~ (UI kit is not ready)
  - [x] Forms submission
  - [x] Forms validation
  - [x] Parameters update
  - [x] Error Handling and permissions
  - [x] Re-introduced hooks 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/8365)
<!-- Reviewable:end -->

mickaelandrieu committed:
BO: started Administration page migration

mickaelandrieu committed:
BO: Migrate administration page to Symfony

mickaelandrieu committed:
BO: fixed display of memcache servers

mickaelandrieu committed:
BO: complete Sf forms

mickaelandrieu committed:
BO: complete Symfony forms

mickaelandrieu committed:
BO: introduced form provider/handler

mickaelandrieu committed:
BO: Added form providers/handlers

mickaelandrieu committed:
BO: Added configuration for form handler

mickaelandrieu committed:
BO: form rendering OK

mickaelandrieu committed:
BO: Form submission works as expected

mickaelandrieu committed:
BO: form submit ok

mickaelandrieu committed:
BO: fixed Codacy warning

mickaelandrieu committed:
BO: mapped GeneralForm

mickaelandrieu committed:
BO: mapped GeneralForm

mickaelandrieu committed:
BO: mapped UploadQuota form

mickaelandrieu committed:
BO: mapped UploadQuota form

mickaelandrieu committed:
BO: mapped Notifications form

mickaelandrieu committed:
BO: mapped Notifications form

mickaelandrieu committed:
BO: changed url to new Controller + delete old one

mickaelandrieu committed:
BO: improved form category logos

mickaelandrieu committed:
TE: remove useless test

Quetzacoalt91 commented:
@mickaelandrieu, it seems this PR also includes #8279. This makes it quite big to be reviewed. :)

Quetzacoalt91 commented:
Accessing the page works properly, thanks. However, submitting new options doesn't work, as the old values keep appeared once the page refreshed.

Quetzacoalt91 commented:
Another request regarding these pages on Firefox, can you please add the class `custom-select` in order to improve the style?

We would get the same result as the first select:
![capture du 2017-10-13 16-49-59](https://user-images.githubusercontent.com/6768917/31554935-1354564e-b037-11e7-81f5-7228bbf229c7.png)

mickaelandrieu commented:
> @mickaelandrieu, it seems this PR also includes #8279. This makes it quite big to be reviewed. :)
I couldnt start this one without features included in #8279, if you merge it... ;)

LouiseBonnard commented on src/PrestaShopBundle/Controller/Admin/FrameworkBundleAdminController.php:
```diff
@@ -228,4 +227,30 @@ protected function authorizationLevel($controller)
+    protected function flashErrors(array $errorMessages)
+    {
+        foreach ($errorMessages as $error) {
+            $this->addFlash('error', $this->trans($error['key'], $error['domain'], $error['parameters']));
```
Normal que key/error/parameters soient entre crochets ?

LouiseBonnard commented on src/PrestaShopBundle/Form/Admin/Type/TranslatorAwareType.php:
```diff
@@ -0,0 +1,58 @@
+     */
+    protected function trans($key, $domain, $parameters = array())
+    {
+        return $this->translator->trans($key, $parameters, $domain);
```
Idem ?

LouiseBonnard commented on src/PrestaShopBundle/Controller/Admin/AdvancedParameters/MemcacheServerController.php:
```diff
@@ -0,0 +1,158 @@
+            )
+        )) {
+            return new JsonResponse(array(
+                'errors' => array($this->trans('You do not have permission to create this.', 'Admin.Notifications.Error')
```
"You do not have permission to add this" ?

LouiseBonnard commented on src/PrestaShopBundle/Controller/Admin/AdvancedParameters/AdministrationController.php:
```diff
@@ -0,0 +1,111 @@
+                PageVoter::LEVEL_DELETE,
+            )
+        )) {
+            $this->addFlash('error', $this->trans('You do not have permission to update this.', 'Admin.Notifications.Error'));
```
"You do not have permission to edit this" ?

mickaelandrieu replied:
be careful, this pull request MUST not create new translation keys: can you confirm the key already exists?

LouiseBonnard replied:
Indeed! And "You do not have permission to update this" does not while "You do not have permission to edit this" already exists in Crowdin.

mickaelandrieu replied:
ok then :+1:

mickaelandrieu committed:
BO: removed old controller

mickaelandrieu committed:
CO: updated routing

mickaelandrieu committed:
CO: coding styles

mickaelandrieu committed:
BO: use custom-select class for select inputs

mickaelandrieu commented:
Comments have been adressed, PR rebased and Coding styles issues fixed => Ready for review :)

mickaelandrieu committed:
TE: removed test on deleted controller

Quetzacoalt91 commented on src/PrestaShopBundle/Controller/Admin/AdvancedParameters/AdministrationController.php:
```diff
@@ -0,0 +1,111 @@
+            $saveErrors = $this->get('prestashop.adapter.administration.form_handler')->save($data);
+
+            if (0 === count($saveErrors)) {
+                $this->addFlash('success', $this->trans('Successful update.', 'Admin.Notifications.Success'));
```
Shouldn't it be "Update successful"? We have both options in the project, what is the best one? (ping @LouiseBonnard)

LouiseBonnard replied:
You're right, both options are available on Crowdin and here it should be "Update successful" because it is a basic notification (context : admin-dev/themes/default/template/controllers/access/helpers/form/form.tpl) when "Successful update" stands for a module (context : modules/autoupgrade/AdminSelfTab.php), thanks !

Quetzacoalt91 commented on src/PrestaShopBundle/Form/Admin/AdvancedParameters/Administration/GeneralType.php:
```diff
@@ -0,0 +1,84 @@
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add('check_modules_update', 'Symfony\Component\Form\Extension\Core\Type\ChoiceType', array(
```
If we confirm we upgrade the minimum PHP version with PS 1.7.4.0, we could use the `::class` constant at last. :)

mickaelandrieu replied:
depends if we merge Symfony3 migration first, which I hope tbh

Quetzacoalt91 reviewed (dismissed):
:ok_hand:

mickaelandrieu commented:
I'll fix comments and rebase after merge of Symfony 3 migration ;)

mickaelandrieu committed:
BO: added custom classes

mickaelandrieu committed:
BO: fixed translation keys

mickaelandrieu committed:
BO: use ::class sugar in forms

mickaelandrieu commented:
Hi @LouiseBonnard keys have been modified, thanks for review :)

@Quetzacoalt91 I've adressed your comments too :+1: 

Can someone from QA team take a look?

Mickaël

Quetzacoalt91 commented:
![capture du 2017-11-20 17-29-18](https://user-images.githubusercontent.com/6768917/33032191-688e355e-ce18-11e7-83c3-ac6aa20fc7d4.png)

I was trying this new page. I could not change the notification options because of that error.

But this quite unclear. What is my actual maximum value? And what is the unit? The information displayed can be improved. :)

mickaelandrieu commented:
> But this quite unclear. What is my actual maximum value? And what is the unit? The information displayed can be improved. :) 

I agree, but it's an "iso-functional" migration :) I still need a specific input from UI kit to display the unit, it's a very specific input type.

Quetzacoalt91 commented:
Why not relying on the default bootstrap 4 display?

This seems to work, and any further change on the ui kit will be already taken in account in the form. :)
![capture du 2017-11-22 15-46-20](https://user-images.githubusercontent.com/6768917/33136170-7f8051ae-cf9c-11e7-8298-d165ba489586.png)

mickaelandrieu committed:
BO: introduct new TextWithUnit form type

mickaelandrieu commented:
@Quetzacoalt91 done!

![done](https://user-images.githubusercontent.com/1247388/33183536-8563f2a6-d078-11e7-8cb5-ce42b6f962e4.png)

Quetzacoalt91 reviewed (dismissed):
oh boiii :100:

marionf commented:
@mickaelandrieu 

In 1.7.2.4 there was tooltips in Administration page

![tooltip administration 1724](https://user-images.githubusercontent.com/13449658/33213479-07d1df00-d128-11e7-89ac-90cf103be269.png)

I don't see them anymore 

![missing tooltip administration 1730](https://user-images.githubusercontent.com/13449658/33213521-2b70b134-d128-11e7-8642-6173dc09c2d1.png)

mickaelandrieu committed:
BO: added all help messages

mickaelandrieu commented:
@marionf sounds better now :+1: 

![admin](https://user-images.githubusercontent.com/1247388/33273731-7877f628-d38e-11e7-9b8b-50b04679b727.png)

Quetzacoalt91 commented:
@mickaelandrieu you sent a screenshot of the new form (with units), but you did not commit them. :(

mickaelandrieu committed:
BO: re-introduced missing specific unity fields

* 'migration/administration-page' of https://github.com/mickaelandrieu/PrestaShop:
  BO: added all help messages
  BO: use ::class sugar in forms
  BO: fixed translation keys
  BO: added custom classes
  TE: remove useless test
  BO: improved form category logos
  BO: changed url to new Controller + delete old one
  BO: mapped Notifications form
  BO: mapped UploadQuota form
  BO: mapped GeneralForm
  BO: form submit ok
  BO: form rendering OK
  BO: introduced form provider/handler
  BO: complete Sf forms
  BO: started Administration page migration
  BO: fixed Codacy warning
  BO: fixed display of memcache servers

mickaelandrieu commented:
@Quetzacoalt91 it happens I spend all my days switching from branches and from Symfony 2 to Symfony 3. Let's release 1.7.3 asap, I have some sort of fatigue of this transitional period :/

![done](https://user-images.githubusercontent.com/1247388/33302527-12904a3a-d3fd-11e7-8751-f483e6bfd2d3.png)

mickaelandrieu commented:
Note that some translations keys have been changed of domain, because this comes from "hint" message of old forms and shouldn't be located into ``Admin.Advparameters.Feature`` domain. It's a bug, this is why I couldn't stay iso-functional this time :+1:

Quetzacoalt91 commented:
Thank you @mickaelandrieu

mickaelandrieu commented:
\o/
